torch/csrc/autograd/engine.cpp

```diff
@@ -342,7 +358,7 @@ auto Engine::thread_main(GraphTask *graph_task) -> void {
     }
   }
 
-  if (graph_task) {
+  if (graph_task && current_depth == 0) {
```

Comment:
This line made me have to pause and think for a moment. It would have been a bit clearer if it said something in a comment like, "If `current_depth == 0`, then this worker thread is done working, and we need to suspend it and transfer control back to the original thread."

This also makes me wonder about an edge case, where you can trigger pathological behavior if you repeatedly make a reentrant call and then exit *right* on one of these "stack chunk" boundaries. Nothing to fix in this PR, but might be worth noting down somewhere in code.